endless looping GIFs with PillowWriter

t-makaro opened the pull request:
## PR Summary

Closes #11787 . ~~Adds an optional parameter to PillowWriter that will set the number of times that a GIF will loop for. Defaults to 0 meaning endless looping. Previously defaulted to 1 (never looping) without a way to change it.~~ API change removed.

Makes PillowWriter produce endless looping gifs like ImageMagickWriter and ImageMagickFileWriter.

## PR Checklist

- ~~[ ] Has Pytest style unit tests~~
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- ~~[ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way~~
- [x] API consistency 

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

t-makaro commented:
Checklist 1: I have no idea how I would create a pytest for this.

Checklist 3,4: I want to add the loop parameter to the PillowWrite.__init__ docstring, but I believe that I would have to copy the entire docstring from MovieWriter.__init__, and then add the extra parameter, so that the sphinx-autodoc will work properly. That's that best way I can think of doing it, is it?

Checklist 5: It's a minor new feature. So, I shouldn't add this?
Checklist 6: It is an api change, but it shouldn't break anything? Like who relied on a gif that only looped once?

jklymak commented:
1. can you interrogate the gif header for the loop parameter?  Probably not so important for a test here.  

3/4 Maybe just in the MovieWriter doc string say that loop exists if the writer is Pillow.  For triple bonus points you could look into the other writers and see if there is a loop method for those.  
5/6 I think a whats new entry is good, but no need for an API entry - thats generally only if API is changed, not if some is added.

WeatherGod commented:
In some respects, this could be considered a bugfix. The original gif
writer that uses ImageMagick produces endless loop GIFs. I don't think
there is an API parameter for that writer as well.

On Sat, Jul 28, 2018 at 9:34 AM, Jody Klymak <notifications@github.com>
wrote:

>
>    1. can you interrogate the gif header for the loop parameter? Probably
>    not so important for a test here.
>
> 3/4 Maybe just in the MovieWriter doc string say that loop exists if the
> writer is Pillow. For triple bonus points you could look into the other
> writers and see if there is a loop method for those.
> 5/6 I think a whats new entry is good, but no need for an API entry -
> thats generally only if API is changed, not if some is added.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/11789#issuecomment-408608355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-EvtAIRCtbVQ0avD7vZOmx1C2g9Vks5uLGhZgaJpZM4Vk6qM>
> .
>

jklymak commented:
Oh, well in that case it’d be good to quickly go through the writers and document their behaviour.

t-makaro commented:
I thought that I'd take a crack a setting the number of loops with ImageMagickWriter. I attempted to modify [this line](https://github.com/matplotlib/matplotlib/blob/6c9c844b7f4ef7b70a232c2e481061102dd5d928/lib/matplotlib/animation.py#L773), but the gif still looped infinitely despite [this doc](https://www.imagemagick.org/script/command-line-options.php#loop). I'll stick with just modifying PillowWriter.

t-makaro commented:
```
Warning, treated as error:
/home/circleci/project/doc/users/next_whats_new/endless_gif.rst:document isn't included in any toctree
```
I think this is just since the next_whats_new toctree is commented out in ```whats_new.rst```

anntzer commented on lib/matplotlib/animation.py:
```diff
@@ -557,6 +557,7 @@ def isAvailable(cls):
     def __init__(self, *args, **kwargs):
```
`__init__(self, *args, loop=False, **kwargs)`
or at least make it a boolean.

Argument needs to be documented in some docstring.

Please check for API consistency across writers; if the behavior of other writers is to loop then the default should be changed here.

t-makaro replied:
The Pillow image.save function takes the ```loop``` parameter as an integer, so a gif could be make to loop a set number of times. Also, ```loop=False``` might make sense in the docstring, but ```loop=self.loop``` where ```self.loop``` is ```False``` is an endless loop since ```False==0```. I suppose I could change it to ```loop=not self.loop```, but should we not allow setting to loop say 5 times? 

I'll change it to a Boolean if you still want. (I suppose no one needs a gif to loop exactly 137 times!)

I like the move to the ```loop``` kwarg in the ```__init__``` function definition. I'll do that for sure.

HTMLWriter and ImageMagick both default to looping I believe, but the ffmpegwriter writing to a .mp4 wouldn't loop. Whether defaulting to loop or not really depends on the format. Gif's should loop, mp4's can't. Different formats have different looping properties (HTMLWriter has a 'reflect' option, so it has a ```default_mode``` parameter that takes a string instead of a ```loop``` parameter. 

I'll default it to looping (unless you object).

I know that I need to put it in a docstring, but I'm having difficulty figuring out where. I'm not familiar with sphinx-autodoc. I really could use help here. I could write an entire new docstring based off the base class and add the parameter. Or I could do as @jklymak suggested, but then other writers would have a "loop parameter only available in PillowWriter".

anntzer commented on doc/users/next_whats_new/endless_gif.rst:
```diff
@@ -0,0 +1,7 @@
+Looping GIFs with PillowWriter
```
Something went wrong with the formatting.

t-makaro replied:
I think the file has windows file endings. I will change that.

t-makaro replied:
I think I changed the line endings?

jklymak commented:
I think you could accept a Boolean and convert to an integer internally; just check if a Boolean.  OTOH it’s a bit confusing that 0 is loop and 1 is loop only once.  Agree that specifying the number loops is desiarable option

t-makaro commented:
I suppose I could do
```
self.loop = not loop if isintance(loop, bool) else loop
```
then set the default to ```loop=True```. Then in whatever docstring say that it accepts a Boolean or an integer that specifies number of loops. This way it is clear that True is endless, False is once, and an integer >=1 is an integer.

jklymak commented:
More that I think about it maybe just make it an integer and value error on anything else so the Boolean doesn’t get passed by accident.  Maybe just lay out the options here and folks can vote 🗳 (I’d do it but am on phone). Sorry to drag out a simple change but may as well get it right.

t-makaro commented:
**Option one:**
Booleans only.
Pros:
- Simple.

Cons:
- can't set a finite >= 1 integer of loops to display.

Implemented as
```python
self.loop = not loop
```

**Option 2:**
ints (this is what it currently is, but without the ValueError)
Pros:
- Can set a finite >= 1 integer of loops to display.

Cons:
- 0 == infinity

Implemented as
```python
# can't use isinstance since bools are ints
if type(loop, int):
    self.loop = loop
else:
    raise ValueError("loop must be an int")
```

**option 3:**
Both bools and ints
Pros:
- simple True is endless and False plays once
- finite number of loops is possible

Cons:
- We have to explain that both are excepted.

Implemented as
```python
self.loop = not loop if isinstance(loop, bool) else loop
```

Documentation will be updated to reflect whatever choice.
I'm in favour of option 2 or 3.

jklymak commented:
I vote 2 + ValueError, just because I think loop=False will actually pass loop=0 if its not caught, but I may be wrong.

t-makaro commented:
I added the ValueError. If we did want to support both bools and ints in the future, then it is easier to add that, then it is to remove support for bools after the fact.

t-makaro commented on lib/matplotlib/animation.py:
```diff
@@ -554,7 +554,38 @@ def isAvailable(cls):
+            output file. Some keys that may be of use include:
+            title, artist, genre, subject, copyright, srcform, comment.
+        '''
+        if type(loop) == int:
```
bools are ints. So ```not isinstance(True, int)``` will return false. Which will result in passing a value of ```True``` aka ```1``` to loop which won't loop, but one would expect loop=True to loop. See my option 3 for a way to implement support for both bools and ints, but this will force just ```ints```.

t-makaro replied:
Though I suppose
```python
if type(loop) != int:
    raise ValueError(...)
self.loop=loop
```
But it only saves one of code, and I don't really feel like pushing another commit to this unless I have too 😛.

t-makaro commented:
That's not a bad point. But I think that the plural ```loops``` makes the fact that a value of 0 results in endless looping make even less sense. Pillow uses the parameter name ```loop```, maybe we should just stick with that.

fredrik-1 commented:
I am looking through the animation module at the moment and based on the available options in all writers at the moment I would say that we could just add unlimited looping also in the pillow writer without any option to change it. Most people want it anyway (but I am not against the option).

I think that a refractoring or at least change in documentation about kwargs should be done in the animation module. At the moment the parameters to the pillow writer are something like, 
``` codec, fps, bitrate, extra_args, metadata``` but the only one that are not ignored is `fps`.

So a new API or documentation that clearly show what options that are used in a writer and adding new relevant options should be good. 

I have started to change some of the documentation to show which kwargs that are actually used and how.

I would suggest to just add the unlimited looping functionality in a pr and the rest in other prs. Do this really need a test?

fredrik-1 commented on lib/matplotlib/animation.py:
```diff
@@ -554,7 +554,38 @@ def isAvailable(cls):
+        loop: int
+            The number of times that the gif will loop.
+            A value of 0 is endless.
+        codec: string or None, optional
```
The rest of the options are just ignored in the pillow writer so I don't think that they should be documented.

fredrik-1 commented:
> I thought that I'd take a crack a setting the number of loops with ImageMagickWriter. I attempted to modify this line, but the gif still looped infinitely despite this doc. I'll stick with just modifying PillowWriter.

Changing that line worked for me. I think that both the pillow- and imagemagick writer API should be the same.

One possibility that are not perfect but is in accordance with the current implementation is to put the new kwarg in `moviewriter` and just put better documentation in the subclasses which kwargs that are used in that writer.

When both imagemagick and pillow seems to use ints for number of loops with 0 as infinity I think that we should stick to that notation.

t-makaro commented:
Sorry it's been a while.

In the interest in getting this bug fix done (and leaving me time for my thousand other projects), I've removed all API changes. This PR now does 2 things:
1. Makes gifs with PillowWriter loop endlessly. Consistent with ```ImageMagickWriter``` and ```ImageMagickFileWriter```.
2. Adds a simple docstring to PillowWriter.

I'll leave any refractoring/API changes to someone more familiar with this API than me. This PR is now the minimum needed to fix #11789. The suggested ```loop``` keyword argument can be added later.

anntzer commented on lib/matplotlib/animation.py:
```diff
@@ -555,6 +555,13 @@ def isAvailable(cls):
+        ----------
+        fps: int
+            Framerate for the gif.
+        '''
```
kill the docstring as well for now?

t-makaro replied:
Sure.

anntzer commented on doc/users/next_whats_new/endless_gif.rst:
```diff
@@ -0,0 +1,7 @@
+Endless Looping GIFs with PillowWriter
+--------------------------------------
+
+We acknowledge that most people want to watch a gif more than once.  Saving an animatation
```
typo (animatation)

t-makaro replied:
Silly me. Fixed.

t-makaro commented:
Ok, I fixed the typo. I believe the doc failures are due to sphinx 1.8.0.

QuLogic commented:
We should have fixed the doc builds now; can you rebase to see if everything's working again?

t-makaro commented:
Last time I tried to rebase a fork, I really messed up the commit history. Any advice?

QuLogic commented:
This is some redundant branch switching:
> ```
> git checkout -b backup-of-master # this will save your bacon if you mess up.
> git checkout master # get back to the changes you made in this PR
> ```

`git branch backup-of-master` will create a branch and not switch to it.

t-makaro commented:
@jklymak Your advice seems to have worked. Thank you. I've rebased.

dstansby commented:
Can whoever merges this squash-merge it?
